### PR TITLE
Added discount field and _validate_discount method to ReceiptEntry.

### PR DESCRIPTION
**Pull request by Alvezgr**
Added discount field for a receipt entry, that is applied to tax base.
e.g: 
```python 
quantity = 2
unit_price = 2
discount = 1
# tax base or total before discount
total = quantity * unit_price
# total = 4 -> 2 * 2 
# discount applied to total
total_price =  quantity * unit_price - discount
# total_price = 3 -> 2 * 2 - 1
```
`discount` is being validated before saving the model due to that negative or greater than total before discount values cannot  be assign to this entry, see: https://github.com/WhyNotHugo/django-afip/blob/aebb5cc45a97143628d9581da53aa84df6b3e2fe/django_afip/models.py#L1512-L1534
Also some tests cases were added, see: https://github.com/WhyNotHugo/django-afip/blob/aebb5cc45a97143628d9581da53aa84df6b3e2fe/tests/test_models.py#L348-L389
closes: #133

### CODE REVIEW

**Comment by WhyNotHugo** [comment]
Sorry for the slowness looking at this. I need to double-check a few details with respect to how AFIP usually models discounts. A bit packed this week, I hope I'll have to to look at this near the end of the week.

**Comment by Alvezgr** [comment]
HI @WhyNotHugo don't worry about that, I understand that this is better to get it right.
To add some hints of what I did, my understanding is that with `wsfev1` AFIP does not models any kind of discount per invoice, what AFIP really cares is your balance (purchases vs sales). I search for it in the `wsfev1` developers manual and I did not found any reference to discount, what I do know is that if you must send invoices via `wsmtxca` service you have to specify every detail even discounts.

I hope this add some clarification and I hope that I'm not too wrong in this.
The reason for making discount direct to `quantity * unit_price` was that I thought that (example with discounts as percentage here, not as net discount as in the model):
AFIP required some percentage of our `net_taxed`.
Let's say that:
`net_taxed = total_per_entry = quantity * unit_price = $83.80` --> then if VAT is 21% we get that
`net_taxed + 21% VAT($17.6) = $101.4`
if we make a discount of 10% to our entry we get that `total_per_entry = $75.42` --> so, now our
`net_taxed = $75.42` 
`net_taxed + 21%VAT($15.84) = $91.26`
that is the same of making the discount to the total of our `net_taxed + VAT`
`net_taxed + 21% VAT - discount = $101.4 - 10% = $91.26`
As we are getting the total VAT calculation latter, we only take into account `net_taxed  = quantity * unit_price - discount`

File: django_afip/models.py [hunk 1]
```diff
@@ -1493,6 +1493,13 @@ class ReceiptEntry(models.Model):
         decimal_places=2,
         help_text=_("Price per unit before vat or taxes."),
     )
+    discount = models.DecimalField(
+        _("discount"),
+        max_digits=15,
+        decimal_places=2,
+        default=0,
+        help_text=_("Total net discount applied to tax base."),
```
Comment (WhyNotHugo) [inline django_afip/models.py:1501]: We can include a validator here to make sure it's positive:

```suggestion
        help_text=_("Total net discount applied to tax base."),
        validators=[MinValueValidator(Decimal("0"))]
```

The relevant import is:

```py
from django.core.validators import MinValueValidator
```

File: django_afip/models.py [hunk 2]
```diff
@@ -1502,10 +1509,42 @@ class ReceiptEntry(models.Model):
...
+        - `discount` must be less than or equal to total price because if discount is
+            greater than total price before discount that will cause total_price
+            to be negative.
+
+        if this validation check fails, a DjangoAfipException will be raised.
+        """
+
+        if self.discount > self.quantity * self.unit_price:
+            msg = "discount must be less than or equal to total price before discount"
+            raise exceptions.DjangoAfipException(_(msg))
```
Comment (WhyNotHugo) [inline django_afip/models.py:1530]: I think it's best to use a [constraint check](https://docs.djangoproject.com/en/4.1/ref/models/constraints/#check) to validate this, so the DB rejects invalid values. E.g.:

```py
CheckConstraint(
    check=Q(discount__lte=F("quantity") * F("unit_price")),
    name='discount_less_than_total'
) 
```

I haven't tried this though; if it has issues we can postpone it and leave this check for now.

File: tests/test_models.py [hunk 3]
```diff
@@ -327,3 +327,63 @@ def test_populate_method(live_ticket):
...
     assert models.CurrencyType.objects.count() == 50
+
+
+@pytest.mark.django_db
+def test_receipt_entry_without_discount():
+    """
+    Test ReceiptEntry.
+
+    Esures that total_price for a ReceiptEntry without a discount
+    works correctly.
```
Comment (WhyNotHugo) [inline tests/test_models.py:338]:
```suggestion
    Ensures that total_price for a ReceiptEntry without a discount
    works correctly.
```

File: tests/test_models.py [hunk 4]
```diff
@@ -327,3 +327,63 @@ def test_populate_method(live_ticket):
...
+    )
+    assert receipt_entry.total_price == 50
+
+
+@pytest.mark.django_db
+def test_receipt_entry_with_valid_discount():
+    """
+    Test ReceiptEntry.
+
+    Esures that total_price for a ReceiptEntry with a valid
```
Comment (WhyNotHugo) [inline tests/test_models.py:353]:
```suggestion
    Ensures that total_price for a ReceiptEntry with a valid
```

File: tests/test_models.py [hunk 5]
```diff
@@ -327,3 +327,63 @@ def test_populate_method(live_ticket):
...
+    with pytest.raises(Exception, match="discount cannot be a negative value"):
+        factories.ReceiptEntryFactory(quantity=5, unit_price=10, discount=-3)
+
+
+@pytest.mark.django_db
+def test_receipt_entry_gt_total_discount():
+    """
+    Test ReceiptEntry discount greater than total price.
+
+    Esures that attempting to generate a ReceiptEntry with a discount
```
Comment (WhyNotHugo) [inline tests/test_models.py:381]:
```suggestion
    Ensures that attempting to generate a ReceiptEntry with a discount
```

**Review by WhyNotHugo — COMMENTED** [review]
Sorry for the delay; I've been travelling recently, but also wanted to double-check some details on how AFIP handles discounts.

I checked a few details on which I had doubts, and I think all looks good. The general idea is close to what AFIP does on their web UI so it all makes sense.

I have minor comments on the validation part mostly. I think we might be able to use DB constraints to handle this for us -- this would be a bit safer, especially in cases where `ReceiptEntry.objects. ... .update(discount=X)` is used.

**Comment by Alvezgr** [comment]
Yes! it makes much more sense the way you put it, using DB constraint seems much better. I'll test everything, commit this changes and update the PR.

PD: Sorry for the typos, already fixed them.

**Pull request closed without merging**